### PR TITLE
Final Submission

### PR DESCRIPTION
**Pull request by CaseyH19**
Casey Hittinger
43199706
I was curious about your last comment saying i did not have enough elixir logic when the game folder has a couple hundred lines of elixir code. Anyways, here is the final product.

### CODE REVIEW

File: connect_4/lib/connect_4_web/channels/room_channel.ex [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    { :ok, socket }
+  end
+
+  def handle_in("turn_played", msg, socket) do
+    game = socket.assigns[:game];
+    g = Game.play(game, msg);
+    socket =
+      socket
+      |> assign(:game, g)
+    broadcast!(socket, "echo", g)
```
Comment (pragdave) [inline connect_4/lib/connect_4_web/channels/room_channel.ex:19]: The use of broadcast here is actually a bug, as it broadcasts the game state to all connected browsers. (Try running the game in two browser windows and making a move in one of them)

# -4

**Comment by pragdave** [comment]
I suspect my previous comment was because I assumed the game logic would be in game.ex, and not state.ex. Normally the state module would be reserved for just the state definition.

I don't see any tests, so -10 for those.

I also don't really see any servers/supervisors. I see that you've copied in the Ex02 code, which does spawn agents, but that's a little dubious in terms of meeting the requirement, so -7 there.

Inside `state.ex` you use agents to simulate variables. We really don't want to use variables in FP. So, for example, the code to divide the board currently looks like this:

~~~ elixir
  def div_boardVerta(count, board) do

    Ex02.reset(count)
    c00 = Enum.filter board, fn _item ->
      Ex02.next_value(count)
      rem(Ex02.get_curr(count), 6) == 0
    end

    Ex02.reset(count)
    c01 = Enum.filter board, fn _item ->
      Ex02.next_value(count)
      rem(Ex02.get_curr(count), 6) == 1
    end

    Ex02.reset(count)
    c02 = Enum.filter board, fn _item ->
      Ex02.next_value(count)
      rem(Ex02.get_curr(count), 6) == 2
    end

    Ex02.reset(count)
    c03 = Enum.filter board, fn _item ->
      Ex02.next_value(count)
      rem(Ex02.get_curr(count), 6) == 3
    end

    Ex02.reset(count)
    c04 = Enum.filter board, fn _item ->
      Ex02.next_value(count)
      rem(Ex02.get_curr(count), 6) == 4
    end

    Ex02.reset(count)
    c05 = Enum.filter board, fn _item ->
      Ex02.next_value(count)
      rem(Ex02.get_curr(count), 6) == 5
    end

    [c01, c02, c03, c04, c05, c00]
  end
~~~

Without variables, and using a function style, it could be

~~~ elixir
  def div_boardVert(_count, board) do
    board
    |> Enum.zip(0..23)
    |> Enum.group_by(fn {_cell, number} -> rem(number, 6) end,
                     fn {cell, _number} -> cell end)
    |> Enum.sort_by(fn {col, _cells} -> col end)
    |> Enum.map(fn {_col, cells} -> cells end)
  end
~~~

This is a really fundamental part of the course: functional programming is about transforming data, and not writing imperative code. I'm deducting 10 style points here.

**Comment by CaseyH19** [comment]
Hello,
My tests were under game/test/game_test.exes
There are multiple tests there.

**Comment by pragdave** [comment]
You are correct. I've added 5 to your grade, bringing it up to 74%. This
does not change the overall grade for the class, which was 76.3%.

Regards


Dave Thomas

**Pull request closed without merging**